src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -102,6 +105,8 @@ namespace ErrorCodes
     extern const int ACCESS_DENIED;
 }
 
+std::unordered_map<IAST::Hash, Data, ASTHash> InterpreterSelectQuery::cached_data;
```

Comment:
std::unordered_map is slow, and also we have predefined cache, check `UncompressedCache.h` and try to reuse it.